### PR TITLE
Add sets section

### PR DESCRIPTION
**Pull request by hargup**
TODO:
- [x] Give brief explanation about various atomic set classes
- [x] Give brief explanation of various set operations.
- [x] Explain how application of set operations on atomic set classes leads to wide range of sets that can be represented by our module.

### CODE REVIEW

File: sets.tex [hunk 1]
```diff
@@ -0,0 +1,84 @@
+%% Sets
+
+SymPy supports representation of a wide variety of set, this is achieved by
```
Comment (ashutoshsaboo) [inline sets.tex]: wide variety of sets

File: sets.tex [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+first defining abstract representation for a smaller number of atomic set
+classes and then combining and transforming them using various set operations.
+
+Each of the set class inherits from the base Set class and defines
```
Comment (ashutoshsaboo) [inline sets.tex]: Each of the set classes

Comment (ashutoshsaboo) [inline sets.tex]: base set class

File: sets.tex [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+rules to check membership of a SymPy object, to calculate union, intersection,
```
Comment (ashutoshsaboo) [inline sets.tex]: defines its own set of rules

File: sets.tex [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+set different. In cases we are not able to evaluate these
```
Comment (ashutoshsaboo) [inline sets.tex]: union, intersection of different sets
In case, we are not able

File: sets.tex [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+operations to atomic set classes they are represented as abstract unevaluated
```
Comment (ashutoshsaboo) [inline sets.tex:10]: atomic set classes, then they are

File: sets.tex [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    \item \textbf{Universal Set}: Everything is a member of Universal Set.
+Union of Universal Set with any set gives Universal Set and intersection leads
+to the other set itself.
+
+    \item \textbf{Finite Set} is functionally equivalent to python's set
+object. Its members can be any object including strings and other sets
+themselves.
+
+    \item \textfb{Range} implements a range of integers and is defined by
```
Comment (ashutoshsaboo) [inline sets.tex]: \textbf{Range} contains a range of integers

File: sets.tex [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    % it.
+    \item \textbf{Empty Set}: Nothing is a member of Empty Set. Union with
+another set returns the other set and intersection leads to an Empty Set.
+
+    \item \textbf{Universal Set}: Everything is a member of Universal Set.
+Union of Universal Set with any set gives Universal Set and intersection leads
+to the other set itself.
+
+    \item \textbf{Finite Set} is functionally equivalent to python's set
+object. Its members can be any object including strings and other sets
```
Comment (ashutoshsaboo) [inline sets.tex]: any objects

File: sets.tex [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+themselves.
```
Comment (ashutoshsaboo) [inline sets.tex]: themselves as well.

File: sets.tex [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    \item \textfb{Range} implements a range of integers and is defined by
+specifying a start value, an end value and a step size. Range is functionally
+equivalent to python's range except the fact that it accepts infinity at end
+points allowing us to represent infinite ranges.
+
+
+    \item \textbf{Real Interval} is specified by giving the start and end point
+and specifying if it is open or closed in these respective ends. The set of
```
Comment (ashutoshsaboo) [inline sets.tex]: in it's respective ends

File: sets.tex [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+real numbers is represented as a special case of a real interval where the
+start point is negative infinite and the end point is positive infinite.
+
+\end{itemize}
+
+
+%% Operations
+
+Other than unevaluated classes of of Union, Intersection and Set Difference
```
Comment (ashutoshsaboo) [inline sets.tex]: Other than the unevaluated classes of Union, Intersection

File: sets.tex [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+operations, we have following set classes.
+
+\begin{itemize}
+
+    \item Product Set abstractly defines the Cartesian product of two or more
```
Comment (ashutoshsaboo) [inline sets.tex]: \textbf{Product Set}

File: sets.tex [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+example to represent a three dimensional space we simple take the Cartesian
+product of three Real sets.
+
+    \item Image Set represents the set of image a function when applied to a
+particular set. In notation Image Set of a function F w.r.t a set S is \{ F(x)
+| x \in S \} In particular we use Image Set to represent the set of infinite
+solutions from trigonometric equations.
+
+
+    \item Condition Set represents subset of a set who's members which
```
Comment (ashutoshsaboo) [inline sets.tex]: \textbf{Condition Set}

File: sets.tex [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+operations, we have following set classes.
+
+\begin{itemize}
+
+    \item Product Set abstractly defines the Cartesian product of two or more
+sets. Product Set is useful when representing higher dimensional spaces. For
+example to represent a three dimensional space we simple take the Cartesian
+product of three Real sets.
+
+    \item Image Set represents the set of image a function when applied to a
```
Comment (ashutoshsaboo) [inline sets.tex]: \textbf{Image Set}

File: sets.tex [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+particular set. In notation Image Set of a function F w.r.t a set S is \{ F(x)
```
Comment (ashutoshsaboo) [inline sets.tex]: a domain set S

File: sets.tex [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+| x \in S \} In particular we use Image Set to represent the set of infinite
```
Comment (ashutoshsaboo) [inline sets.tex]: {F(x) | x in S}.

File: sets.tex (hunk 12 above)
Comment (ashutoshsaboo) [inline sets.tex]: represents the subset
whose members satisfy a particular

File: sets.tex [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+solutions from trigonometric equations.
+
+
+    \item Condition Set represents subset of a set who's members which
+satisfies a particular condition. In notation Condition Set of set S w.r.t to a
+condition H is \{x | H(x), x \in S \}. We use Condition Set to represent the
```
Comment (ashutoshsaboo) [inline sets.tex]: In notation, the Condition Set of set S w.r.t to a condition H is {x | H(x), x \in S}

File: sets.tex [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+\end{itemize}
+
+
+
+
+
+%% Explaining it later because it is a special case of Image Set rather being
+%% something atomic
+
+    \textbf{Complex Region}
```
Comment (ashutoshsaboo) [inline sets.tex]: So, are you planning to add more details under this `Complex Region`? @hargup If not, then consider removing this heading. 

Also, do the same for the comments you have mentioned below (towards the end of this file). @hargup :smiley:

**Comment by ashutoshsaboo** [comment]
Hi @hargup . I have made several line comments. Look into them, and correct the errors. Also, I would suggest, have a look at the paper once by building the LaTeX into a PDF file on your local system as well using `pdf-latex`. 

Thanks. :smile: 

cc/ @asmeurer

File: sets.tex [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+%% Operations
+
+Other than unevaluated classes of of Union, Intersection and Set Difference
+operations, we have following set classes.
+
+\begin{itemize}
+
+    \item Product Set abstractly defines the Cartesian product of two or more
+sets. Product Set is useful when representing higher dimensional spaces. For
+example to represent a three dimensional space we simple take the Cartesian
```
Comment (gxyd) [inline sets.tex]: simple -> simply

File: sets.tex (hunk 4 above)
Comment (gxyd) [inline sets.tex]: different -> difference

File: sets.tex (hunk 13 above)
Comment (gxyd) [inline sets.tex]: This doesn't sound good.

File: sets.tex (hunk 7 above)
Comment (asmeurer) [inline sets.tex]: It doesn't actually allow strings. I would remove that.

File: sets.tex [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+rules to check membership of a SymPy object, to calculate union, intersection,
+set different. In cases we are not able to evaluate these
+operations to atomic set classes they are represented as abstract unevaluated
+objects.
+
+
+\begin{itemize}
+
+    % Description of Empty Set sounds too obvious, not sure if we need to keep
+    % it.
```
Comment (asmeurer) [inline sets.tex]: You could just write "represents the empty set $\emptyset$".

**Comment by asmeurer** [comment]
Put the name of the SymPy class in each item (like "\verb|EmptySet|").

File: sets.tex (hunk 1 above)
Comment (Upabjojr) [inline sets.tex]: shouldn't _representation_ and _set_ be plural?

File: features.tex [hunk 20]
```diff
@@ -8,6 +8,9 @@ \subsection{Basic Operations}
 % descriptions will go in the algorithms section.
 \subsection{Calculus}
 
+% Sets
+\subsection{Sets}
+
```
Comment (Upabjojr) [inline features.tex:82]: missing _\input{sets}_ ?

**Comment by ashutoshsaboo** [comment]
@hargup Please make the changes as described above^, so that we can consider merging this PR, if everything seems to be fine after that! :smiley:

**Comment by moorepants** [comment]
@hargup Please fix all of the comments ASAP so that we can merge the first draft.

**Comment by hargup** [comment]
@ashutoshsaboo @moorepants I've fixed all the comments. Thanks!

File: sets.tex [hunk 21]
```diff
@@ -0,0 +1,86 @@
+%% Sets
+
+SymPy supports representation of a wide variety of sets, this is achieved by
+first defining abstract representation for a smaller number of atomic set
+classes and then combining and transforming them using various set operations.
+
+Each of the set classes inherit from the base set class and defines rules to
```
Comment (shivamvats) [inline sets.tex:7]: inherits

Comment (ashutoshsaboo) [inline sets.tex:7]: @shivamvats I feel it will be inherit only. Because then it'll be two plurals together, which won't be grammatically correct.

**Comment by ashutoshsaboo** [comment]
@hargup Your Travis tests seem to fail. 

Please see the error log below (https://travis-ci.org/sympy/sympy-paper/builds/124103450) -:

```
LaTeX Warning: No \author given.
(./introduction.tex) (./architecture.tex [1{/var/lib/texmf/fonts/map/pdftex/upd
map/pdftex.map}]
Overfull \hbox (27.99953pt too wide) in paragraph at lines 56--64
[]\OT1/cmr/m/n/10 Every SymPy ex-pres-sion sat-is-fies a key in-vari-ant, namel
y, \OT1/cmtt/m/n/10 expr.func(*expr.args)
[2] [3]
Overfull \hbox (61.97145pt too wide) in paragraph at lines 204--213
\OT1/cmr/m/n/10 sump-tions de-fined, can be rewrit-ten in terms of fac-to-rial 
with \OT1/cmtt/m/n/10 gamma(x).rewrite(factorial)\OT1/cmr/m/n/10 ,
[4]
Overfull \hbox (17.24684pt too wide) in paragraph at lines 245--245
[]            \OT1/cmtt/m/n/10 return self.func(self.args[0])*polygamma(0, self
.args[0])[] 
) (./algorithms.tex
LaTeX Warning: Citation `gruntz1996computing' on page 5 undefined on input line
 19.
Overfull \hbox (27.74675pt too wide) in paragraph at lines 28--28
[]\OT1/cmtt/m/n/10 In [2]: limit((2*E**((1-cos(x))/sin(x))-1)**(sinh(x)/atan(x)
**2), x, 0)[] 
[5]
Overfull \hbox (31.31221pt too wide) in paragraph at lines 54--60
[]\OT1/cmr/m/n/10 Strategy: mrv set: the set of most rapidly vary-ing subex-pre
s-sions $\OMS/cmsy/m/n/10 f\OML/cmm/m/it/10 e[]; e[]; e[]\OMS/cmsy/m/n/10 g$\OT
1/cmr/m/n/10 ,
LaTeX Warning: Citation `gruntz1996computing' on page 6 undefined on input line
 75.
) (./features.tex
LaTeX Warning: Reference `features-table' on page 6 undefined on input line 5.
[6]
Underfull \hbox (badness 1107) in paragraph at lines 16--19
[]|\OT1/cmr/m/n/10 Summations, prod-ucts, bi-no-mial co-ef-fi-cients, prime
LaTeX Warning: Citation `petkovvsek1996bak' on page 7 undefined on input line 2
2.
LaTeX Warning: Citation `Hunter:2007' on page 7 undefined on input line 23.
LaTeX Warning: Citation `adams1994introduction' on page 7 undefined on input li
ne 34.
LaTeX Warning: Citation `Nijenhuis1978combinatorial' on page 7 undefined on inp
ut line 48.
LaTeX Warning: Citation `biggs1976graph' on page 7 undefined on input line 49.
Underfull \hbox (badness 1642) in paragraph at lines 45--50
[]|\OT1/cmr/m/n/10 Implements per-mu-ta-tions, com-bi-na-tions, par-ti-tions,
Overfull \hbox (61.90755pt too wide) in paragraph at lines 12--69
[][] 
LaTeX Warning: Float too large for page by 215.08887pt on input line 69.
(./features_basic_operations.tex [7] [8pdfTeX warning (ext4): destination with 
the same identifier (name{table.1}) has been already used, duplicate ignored
\AtBegShi@Output ...ipout \box \AtBeginShipoutBox 
                                                  \fi \fi 
l.69 
     ]
Overfull \hbox (69.99886pt too wide) in paragraph at lines 96--100
\OT1/cmr/m/n/10 pres-sion through the log-i-cal con-nec-tor class, in this case
 []\OT1/cmtt/m/n/10 And(Q.positive(x), Q.positive(y))\OT1/cmr/m/n/10 . 
[9]) (./sets.tex (/usr/share/texlive/texmf-dist/tex/latex/base/omscmr.fd)
[10]
! Undefined control sequence.
<recently read> \mathbb 

l.28 ...egers| represents set of Integers $\mathbb
                                                  {Z}$.
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on paper.log.
make: *** [travis] Error 1
```

Please fix this, so that we can merge it for moving ahead. @hargup Also, I'd would advice you (if you have latex locally installed) to try building the PDF on your local system as well, before pushing the commit. @hargup

**Comment by ashutoshsaboo** [comment]
@hargup Yes now it seems to be good. I guess it's ready for merge. 

Thanks! :+1: 

cc/ @moorepants @asmeurer 

Also, the master passes the build tests after the merge. So, seems fine I guess. :+1:

File: sets.tex (hunk 21 above)
Comment (asmeurer) [inline sets.tex:7]: "Inherits" is correct. The noun here is "each", which is singular.

**Pull request merged**